// astro-imagetools/0.8.1/astro-imagetools/utils/runtimeChecks.js
// @ts-check
import fs from "node:fs";
import path from "node:path";
import findCacheDir from "find-cache-dir";
import filterConfigs from "0.8.1/astro-imagetools/utils/filterConfigs.js";

const FindUpModule = await import("find-up"),
  findUp = FindUpModule.findUp || FindUpModule.default;

// Sharp related checks
export const sharp = await (async () => {
  try {
    if (await import("sharp")) {
      return true;
    }
  } catch (error) {
    return false;
  }
})();

export const supportedImageTypes = [
  "avif",
  "jpeg",
  "jpg",
  "png",
  "webp",
  ...(sharp ? ["heic", "heif", "tiff", "gif"] : ["jxl", "wp2"]),
];

// prettier-ignore
export const supportedConfigs = [
  "src", "alt", "tag", "content", "sizes", "preload", "loading", "decoding", "attributes",
  "layout", "placeholder", "breakpoints", "objectFit", "objectPosition", "backgroundSize",
  "backgroundPosition", "format", "fallbackFormat", "includeSourceFormat", "formatOptions",
  "fadeInTransition", "artDirectives", "flip", "flop", "invert", "flatten", "normalize",
  "grayscale", "hue", "saturation", "brightness", "w", "h", "ar", "width", "height", "aspect",
  "background", "tint", "blur", "median", "rotate", "quality", "fit", "kernel", "position",
  "cacheDir", "assetFileNames", "globalImportRemoteImage",
];

const configFile = await findUp([
  "astro-imagetools.config.js",
  "astro-imagetools.config.mjs",
]);

const configFunction = configFile
  ? await import(configFile).catch(async () => await import("/" + configFile))
  : null;

const rawGlobalConfigOptions = configFunction?.default ?? {};

const NonGlobalConfigOptions = ["src", "alt", "content"];

const GlobalConfigs = supportedConfigs.filter(
  (key) => !NonGlobalConfigOptions.includes(key)
);

const GlobalConfigOptions = filterConfigs(
  "Global",
  rawGlobalConfigOptions,
  GlobalConfigs
);

if (GlobalConfigOptions.globalImportRemoteImage) {
  const DisclaimerComment = `// Generated by 'astro-imagetools'`,
    GlobalFunctionType = `declare global {\n  function importRemoteImage(url: string): Promise<string>;\n}`;

  if (fs.existsSync("./auto-imports.d.ts")) {
    fs.appendFileSync(
      "./auto-imports.d.ts",
      "\n" + DisclaimerComment + "\n" + GlobalFunctionType + "\n"
    );
  } else {
    fs.writeFileSync(
      "./importRemoteImage.d.ts",
      DisclaimerComment + "\nexport {};\n\n" + GlobalFunctionType + "\n"
    );
  }
}

export { GlobalConfigOptions };

// CWD
export const cwd = process.cwd().split(path.sep).join(path.posix.sep);

const { cacheDir } = GlobalConfigOptions;

// FS Cache related checks
const fsCachePath =
  (cacheDir
    ? cwd + cacheDir
    : findCacheDir({
        name: "astro-imagetools",
      })) + "/";

fs.existsSync(fsCachePath) || fs.mkdirSync(fsCachePath, { recursive: true });

export { fsCachePath };
