// astro-imagetools/0.8.1/astro-imagetools/integration/index.js
// @ts-check
import fs from "node:fs";
import { fileURLToPath } from "node:url";
import { posix as path, resolve } from "node:path";
import { saveAndCopyAsset } from "0.8.1/astro-imagetools/integration/utils/saveAndCopyAsset.js";
import vitePluginAstroImageTools, { store } from "0.8.1/astro-imagetools/plugin";

const filename = fileURLToPath(import.meta.url);

const astroViteConfigsPath = resolve(filename, "../../astroViteConfigs.js");

export default {
  name: "astro-imagetools",
  hooks: {
    "astro:config:setup": async function ({ config, command, updateConfig }) {
      const environment = command;

      const isSsrBuild = command === "build" && !!config.adapter;

      let projectBase = path.normalize(config.base);

      if (projectBase.startsWith("./")) projectBase = projectBase.slice(1);

      if (!projectBase.startsWith("/")) projectBase = "/" + projectBase;

      if (projectBase.endsWith("/")) projectBase = projectBase.slice(0, -1);

      const astroViteConfigs = {
        environment,
        isSsrBuild,
        projectBase,
        publicDir: fileURLToPath(config.publicDir.href),
        rootDir: fileURLToPath(config.root.href),
      };

      await fs.promises.writeFile(
        astroViteConfigsPath,
        `export default ${JSON.stringify(astroViteConfigs)}`
      );

      updateConfig({
        vite: {
          plugins: [vitePluginAstroImageTools],
        },
      });
    },

    "astro:build:done": async function closeBundle() {
      const { default: astroViteConfigs } = await import(
        // @ts-ignore
        "../astroViteConfigs.js"
      );

      const { mode, outDir, assetsDir, isSsrBuild } = astroViteConfigs;

      if (mode === "production") {
        const allEntries = [...store.entries()];

        const assetPaths = allEntries.filter(
          ([, { hash = null } = {}]) => hash
        );

        await Promise.all(
          assetPaths.map(
            async ([assetPath, { hash, image, buffer }]) =>
              await saveAndCopyAsset(
                hash,
                image,
                buffer,
                outDir,
                assetsDir,
                assetPath,
                isSsrBuild
              )
          )
        );
      }
    },
  },
};
